File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        if not nums:
            return []
    
        res = []
    
        def dfs(nums: List[int], curr):
            if curr is None:
                curr = []
    
            curr_sum = sum(curr) if curr else 0
    
            if curr_sum == target:
                if curr not in res:
                    res.append(curr)
                return

    
            if len(nums) == 0 or curr_sum > target:
                return
    
            dfs(nums[1:].copy(), curr=curr.copy())
    
            curr.append(nums[0])
    
            dfs(nums.copy(), curr=curr.copy())
    
            dfs(nums[1:].copy(), curr=curr.copy())
    
        dfs(nums, curr=[])
    
        return res

    

File: Data Structures & Algorithms/subsets/submission-0.go
func subsets(nums []int) [][]int {
	res := [][]int{}

	return dfs(res, nums, []int{})
}

func dfs(res [][]int, nums []int, curr []int) [][]int {
	if len(nums) <= 0 {
		res = append(res, curr)
		return res
	}
	num := nums[0]
	newCurr := make([]int, len(curr))
	copy(newCurr, curr)
	newCurr = append(newCurr, num)

	res = dfs(res, nums[1:], newCurr)
	res = dfs(res, nums[1:], curr)
	return res
}
